### ISSUE TITLE
Regression: Ordering by joined column doesn't return results

### ISSUE THREAD

**Issue by DDtKey** [apache/arrow-datafusion#8374]
### Describe the bug

After update to `datafusion: 33` I've noticed wrong behavior for our internal test with sorting by multiple columns.
It used to work in `datafusion: 31`

### To Reproduce

MRE with datafusion-cli:

```sql
CREATE TABLE users AS VALUES('Alice',50),('Bob',100);
CREATE TABLE employees AS VALUES('Alice','Finance'),('Bob','Marketing');

SELECT u.* FROM users u JOIN employees e ON u."column1" = e."column1" ORDER BY u."column1", e."column2";
0 rows in set. Query took 0.002 seconds.
```

But at the same time, without ordering by joined column it works:

```sql
SELECT u.* FROM users u JOIN employees e ON u."column1" = e."column1" ORDER BY u."column1";
+---------+---------+
| column1 | column2 |
+---------+---------+
| Alice   | 50      |
| Bob     | 100     |
+---------+---------+
2 rows in set. Query took 0.002 seconds.
```

### Expected behavior

It should work as before

**Comment by DDtKey** [comment]
> When the column names are identical, the error will arise

Just to clarify: in my tests this **failed with different column names** as well. Just MRE uses auto column names

**Comment by haohuaijin** [comment]
> Just to clarify: in my tests this **failed with different column names** as well. Just MRE uses auto column names

@DDtKey could you provide some cases? When the column name is different, I find it works in `datafusion 33` 06bbe1298fa8aa042b6a6462e55b2890969d884a
```
DataFusion CLI v33.0.0
❯ create table u(a text, b int) as values ('Alice', 50), ('Bob', 100);
0 rows in set. Query took 0.023 seconds.

❯ create table e(c text, d text) as values ('Alice', 'Finance'), ('Bob', 'Marketing');
0 rows in set. Query took 0.000 seconds.

❯ select u.* from u join e on u.a = e.c order by u.a, e.d;
+-------+-----+
| a     | b   |
+-------+-----+
| Alice | 50  |
| Bob   | 100 |
+-------+-----+
2 rows in set. Query took 0.021 seconds.
```

**Comment by DDtKey** [comment]
Sorry for the misunderstanding, you're correct
It works with different column names (used in `ORDER BY`), so that seems to be the root of the problem.

**Comment by Asura7969** [comment]
My initial solution：
```rust
.find_map(|(index, (projected_expr, alias))| {
  projected_expr.as_any().downcast_ref::<Column>().and_then(
      |projected_column| {
          (column.index() == projected_column.index()       <--- and index comparison
              && column.name().eq(projected_column.name()))
          .then(|| {
              state = RewriteState::RewrittenValid;
              Arc::new(Column::new(alias, index)) as _
          })
      },
  )
})
```
> and index comparison

```text
DataFusion CLI v33.0.0
❯ CREATE TABLE users AS VALUES('Alice',50),('Bob',100);
0 rows in set. Query took 0.022 seconds.

❯ CREATE TABLE employees AS VALUES('Alice','Finance'),('Bob','Marketing');
0 rows in set. Query took 0.008 seconds.

❯ SELECT u.* FROM users u JOIN employees e ON u."column1" = e."column1" ORDER BY u."column1", e."column2";
+---------+---------+
| column1 | column2 |
+---------+---------+
| Alice   | 50      |
| Bob     | 100     |
+---------+---------+
```
The result is correct

**Comment by Asura7969** [comment]
replenish:
![image](https://github.com/apache/arrow-datafusion/assets/26200914/a0f04df1-4135-42bb-93b1-89e9a18fe56e)
Maybe we should do special processing for `SortExec` 🤔

**Comment by haohuaijin** [comment]
> My initial solution：
> 
> ```rust
> .find_map(|(index, (projected_expr, alias))| {
>   projected_expr.as_any().downcast_ref::<Column>().and_then(
>       |projected_column| {
>           (column.index() == projected_column.index()       <--- and index comparison
>               && column.name().eq(projected_column.name()))
>           .then(|| {
>               state = RewriteState::RewrittenValid;
>               Arc::new(Column::new(alias, index)) as _
>           })
>       },
>   )
> })
> ```

use name and index(the index is column index of input schema) to identify a column, should be under the assumption that the input schema of `column`'s plan and `projection_column`'s plan is the same. Otherwise, some projection that can be pushed down may become unable to be pushed down. And when the schema is the same, we can just use the index to identify a column.

**Comment by DDtKey** [comment]
Why don't we consider this problem a regression and continue to release new stable versions?
This worked until version "31", since then we have "34" and the regression is ignored 🤔

Thus we have the following situation:
Each version needs to code adaptation in case of incompatible changes, but it is not correct to use the new version with obvious regression. Thus, the changes are made but cannot be applied (and their number is growing)

I may be a little behind the current `datafusion` release policy, but I think any regression should be prioritized to release new stable patches.

**note**: I'm not talking about bugs in general, but about regressions, unfortunately they occur quite often and they are more dangerous, there is no trust in new versions

cc @alamb

**Comment by alamb** [comment]
> note: I'm not talking about bugs in general, but about regressions, unfortunately they occur quite often and they are more dangerous, there is no trust in new versions


Thank you for bringing this up -- I agree we need to prioritize regressions -- I personally missed this particular bug as a regression and thought it was a pre-existing bug.  I have updated the title to reflect this and created a new tag for regressions

**Comment by ozankabak** [comment]
@DDtKey I think some people may have assumed #8485 fixed it (at least I did). You are right that such regressions should get priority and we will prioritize this.

**Comment by DDtKey** [comment]
@ozankabak thanks for pointing to the PR. Looks like I've missed that it has been merged prior to releasing `34.0.0` (and the issue has not closed yet). 

So that's my wrong assumption, sorry. I tested and this case works with the latest version. 
Though, as it's been mentioned we may have some underlying issues, but not related to this one.

**Comment by alamb** [comment]
BTW one of the longer term discussions I would like to have at https://github.com/apache/arrow-datafusion/discussions/8152 and in other venues (I just haven't had time to write it down yet) is how to improve the overall "process maturity" of datafusion -- like @DDtKey  points out that regressions should be prioritized, but at the moment we don't really have a mechanism to do that (or, for example, hold the release for such regressions) other than by relying on one of us to catch it manually

**Comment by DDtKey** [comment]
Should we close this issue as fixed in `34.0.0` to avoid confusion?

**Comment by ozankabak** [comment]
@DDtKey sounds good 👍
